apps/src/code-studio/header/MinimalProjectHeader.jsx

```diff
@@ -0,0 +1,67 @@
+/* globals dashboard, appOptions */
+
+import PropTypes from 'prop-types';
+import React from 'react';
+import {connect} from 'react-redux';
+
+// Minimal project header for viewing channel shares and legacy /c/ share pages.
+class MinimalProjectHeader extends React.Component {
+  static propTypes = {
+    isSignedIn: PropTypes.bool
+  };
+
+  state = {
+    currentName: undefined
+  };
+
+  remixProject = () => {
+    if (
+      dashboard.project.getCurrentId() &&
+      dashboard.project.canServerSideRemix()
+    ) {
+      dashboard.project.serverSideRemix();
+    } else if (!this.props.isSignedIn) {
+      window.location = `/users/sign_in?user_return_to=${
+        window.location.pathname
+      }`;
+    } else {
+      // We don't have an id. This implies we are either on a legacy /c/ share
+      // page or a script level. In these cases, copy will create a new project
+      // for us.
+      const newName =
+        'Remix: ' +
+        (dashboard.project.getCurrentName() ||
+          appOptions.level.projectTemplateLevelName ||
+          'My Project');
+      dashboard.project
+        .copy(newName, {shouldNavigate: true})
+        .then(() => this.setState({currentName: newName}))
+        .catch(err => console.log(err));
+    }
+  };
+
+  render() {
+    return (
+      <div>
+        <div className="project_name_wrapper header_text">
+          <div className="project_name header_text">
```

Comment:
rephrasing https://github.com/code-dot-org/code-dot-org/pull/28422#discussion_r283048875 in an actionable way -- I think you will want to use ProtectedStatefulDiv here, to protect header_text (or another element of your choosing), so that non-react code can render into it without worrying about it getting blown away when react re-renders.

this step could be skipped if a few of these PRs/commits can be grouped together in a way that no non-react updates to react DOM elements remain, or if you can convince yourself that react re-renders do not happen on this component for some reason.

Reply:
another option could be to use `shouldComponentUpdate` on this component to prevent a re-render.

Reply:
This is actually precisely why I stopped pursuing this "incremental updates" approach in favor of the monolithic change in https://github.com/code-dot-org/code-dot-org/pull/28461; the control flows in the jquery version were just entirely too tangled for me to try and think about in a react-compatible way.

In this _specific_ case, I actually think we're fine; the only pieces of code that are modifying `.project_name` are the other `showXHeader` methods which aren't used at the same time as the MinimalProjectHeader and the `dashboard.header.remixProject` method, which is replaced in this PR with the `MinimalProjectHeader.remixProject` implementation.

I believe the same is true for the components extracted in the LevelbuilderSaveButton PR, but as soon as I started to work on the other headers I realized that their internal rendering logic was much more complicated, and could not be done without replacing everything all at once.

Reply:
ok, so to make progress here, would it be safe in this sense to merge this PR and the LevelbuilderSaveButton change in https://github.com/code-dot-org/code-dot-org/pull/28422 ? and then at least the rest of this work would be slightly less monolithic? I'd like to break this into chunks if possible, but agree that going down the frankenstein jquery + react path is worth trying hard to avoid.

Reply:
Yeah, that SGTM!